### DISCUSSION TITLE
Which backend technology should be used ?

### DISCUSSION THREAD

**Discussion by grzanka** [yaptide/yaptide#55] [General]
Flask vs django ?

**Comment by Derstilon** [comment]
https://geek.justjoin.it/flask-vs-django-dlaczego-nie-warto-wybierac-flaska-w-2021-roku

https://hackernoon.com/advantages-and-disadvantages-of-django-499b1e20a2c5

**Comment by grzanka** [comment]
Here is a short list of criteria or aspects to consider (random order):
 - ORM 
 - plugins for authentication
 - community size and support
 - ability to handle standalone static UI (as our 3JS editor) + optional backend for logged user
 - popularity among the students and chances to find manpower for future project development and maintenance
 - release schedule and risk (and burden) related to maintenance and migration:
    - how often new releases are published?
    - Is it difficult to migrate to newer versions ?
    - how long old releases are supported (are there any LTS versions)?

**Comment by p1003** [comment]
https://hackr.io/blog/flask-vs-django
https://www.simform.com/blog/flask-vs-django/

**Comment by p1003** [comment]
Django vs Flask
ORM:
 - Both Django ORM and SqlAlchemy are competent solutions and offer sufficient features for our usecase. We slightly prefer SqlAlchemy since it's independent from Flask and therefore more people might have used it in their previous projects. Additionally learning it may become useful in future projects.
 
Authentication:
 - Authentication shouldn't be an issue in any of the two frameworks, since it will be integrated into the frontend part of the app, and the backend will just store and validate secrets. Both frameworks can do that.

UI:
 - We will be communicating with a standalone React front-end application via REST. We won't be serving any UI. Both frameworks are capable of doing that though.

Popularity:
 - Django and Flask have almost the same amount of stars on GitHub, however Flask is more popular on StackOverflow. Since flask is smaller and uses modules for some of its functionalities it should be easier to find someone familiar with at least some of them compared to a person that knows Django well. In any case both frameworks are similar and it shouldn't be a problem to convert from using one over to the other. Subjectively, Flask seems much less intimidating than Django due to its size.

Migration Risk:
 - Since flask uses python modules that are updated independently, we risk more frequent, smaller updates to them when compared with updates to Django as a whole. Django does a major release once a year and Flask has irregular small updates. This section comes down to preference in my opinion.

We are leaning towards Flask mainly due to its accessibility and more online resources. Learning many smaller modules rather than a big framework also seems more compelling to us in terms of our growth and might appeal to our successors.

~Łukasz and Krzysztof

**Reply by grzanka** [reply]
Most of the points look very reasonable. What about the recent (2021-05-11) major migration of Flask to version 2.0 ?
See https://flask.palletsprojects.com/en/2.0.x/changes/ (up to now only minor 2.0.1 was released).

Is there enough posts on StackOverflow on features of 2.x to safely work with it? If not we will be facing problems of reading about some solutions from 1.x (released on 2018-04-26) or 0.x (2010-2018) which cannot be easily adapted to 2.x.
